### ISSUE TITLE
JSON Serialization Name Collision with hidden properties

### ISSUE THREAD

**Issue by douglasg14b** [dotnet/runtime#30964]
Take Two Classes:

```cs
public class ClassA
{
    public List<int> SomeList { get; set; }
}

public class ClassB : ClassA
{
    new public IEnumerable<int> SomeList { get; set; }
}
```

If `ClassB` is serialized an exception is thrown:

>The JSON property name for 'ClassB.SomeList' collides with another property.

This seems like unwanted behavior.

**Comment by layomia** [comment]
The expectation here is that property name collisions due to a member on a parent being hidden (with the `new`) should be resolved by the serializer ignoring the member on the parent. e.g. serializing an instance of `MyDerivedClass` should work fine:

```C#
public class MyClass
{
    // This should be ignored.
    public int MyNumber { get; set; }
}
	
public class MyDerivedClass : MyClass
{
    // This should be (de)serialized.
    new public double MyNumber { get; set; }
}
```

when field support is added, this should also work fine:

```C#
public class MyClass
{
    // This should be ignored.
    public int MyNumber;
}
	
public class MyDerivedClass : MyClass
{
    // This should be (de)serialized.
    new public double MyNumber { get; set; }
}
```

`[JsonIgnore]` should continue to work as expected (same for permutations with fields):

```C#
public class MyClass
{
    // This should be ignored.
    [JsonIgnore]
    public int MyNumber { get; set; }
}
	
public class MyDerivedClass : MyClass
{
    // This should be (de)serialized.
    new public double MyNumber { get; set; }
}
```

```C#
public class MyClass
{
    // This should be (de)serialized.
    public int MyNumber { get; set; }
}
	
public class MyDerivedClass : MyClass
{
    // This should be ignored.
    [JsonIgnore]
    new public double MyNumber { get; set; }
}
```

Property name collisions due to `[JsonPropertyName]` or `JsonNamingPolicy` should continue to fail with the collision error:

```C#
public class MyClass
{
    public int MyNumber { get; set; }
}
	
public class MyDerivedClass : MyClass
{
    [JsonPropertyName("MyNumber")]
    public double MyDouble { get; set; }
}
```

Newtonsoft.Json does not throw a collision error in such cases and honors the configuration on the derived class (ignores the parent member):

```C#
public class MyClass
{
    // This is ignored by Newtonsoft.Json
    public int MyNumber { get; set; }
}
	
public class MyDerivedClass : MyClass
{
    // This is (de)serialized by Newtonsoft.Json
    [JsonProperty("MyNumber")]
    public double MyDouble { get; set; }
}
```

I don't think the presence of a colliding property name (due to `[JsonPropertyInfo]`/`JsonNamingPolicy`) on a member of deriving class is enough information for the serializer to ignore a member on a parent. `[JsonIgnore]` should be used in such cases:

```C#
public class MyClass
{
    // This should be ignored
    [JsonIgnore]
    public int MyNumber { get; set; }
}
	
public class MyDerivedClass : MyClass
{
    // This should be (de)serialized
    [JsonPropertyName("MyNumber")]
    public double MyDouble { get; set; }
}
```

`[JsonPropertyName]`/`JsonNamingPolicy` should work as expected on `new` properties, so long as there are no collisions:

```C#
public class MyClass
{
    // This should be (de)serialized
    public int MyNumber { get; set; }
}
	
public class MyDerivedClass : MyClass
{
    // This should be (de)serialized
    [JsonPropertyName("MyDouble")]
    new public double MyNumber { get; set; }
}
```

cc @steveharter, @ahsonkhan

The implementation to fix this issue should include tests for all such permutations.

**Comment by YohDeadfall** [comment]
I've found another case reported by #32106. An interesting thing happens when you ask for properties of type when it has a new slot member. In case of @douglasg14b reflection returns three properties, but in the example below it returns only one from the derived class.

Is it expected behavior, @GrabYourPitchforks @steveharter ?

```csharp
var case1 = typeof(ClassWithNewSlotPrivateProperty).GetProperties(BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic);
// ClassWithNewSlotPrivateProperty.MyString

var case2 = typeof(ClassB).GetProperties(BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic);
// ClassB.P1
// ClassA.P1
// ClassA.P2
```

```csharp
// Case 1
public class ClassWithPublicProperty
{
    public string MyString { get; set; } = "DefaultValue";
}

public class ClassWithNewSlotPrivateProperty : ClassWithPublicProperty
{
    private new string MyString { get; set; } = "NewDefaultValue";
}


// Case 2
public class ClassA
{
    public List<int> SomeList { get; set; }
}

public class ClassB : ClassA
{
    new public IEnumerable<int> SomeList { get; set; }
}
```

/cc @jkotas

**Comment by NinoFloris** [comment]
@YohDeadfall to fully understand your example, you think case 1 should (de)serialize `ClassWithPublicProperty.MyString` because the derived class new slot property is private, and therefore less visible?

I guess I can agree to having every new slot property with higher or equal visibility being preferred over its parent. Is that the plan? Ultimately it's another thing to keep in mind, always hiding the parent if shadowed irrespective of visibility is simpler. Do you have some examples where it makes sense to go for the more complex logic?

**Comment by YohDeadfall** [comment]
@NinoFloris, sorry for the late reply. Yes, I think that derived classes should not alter serialization of base class members in a way to hide them. Hiding a parent's member by a derived class isn't a good idea since when something is public then there are some sense behind this, and changing visibility to private highlights architecture problems.

**Comment by layomia** [comment]
cc @steveharter, @Jozkee

**Comment by YohDeadfall** [comment]
@layomia this issue should be fixed by #32107. Just need to write one more test and change naming style for new visibility tests as requested.

**Comment by steveharter** [comment]
> I've found another case reported by #32106. An interesting thing happens when you ask for properties of type when it has a new slot member. In case of @douglasg14b reflection returns three properties, but in the example below it returns only one from the derived class.

> Is it expected behavior, @GrabYourPitchforks @steveharter

The `case2` should only return 2 items, not 3.

If the name and signature of a property matches a derived class, [it is considered a duplicate](https://github.com/dotnet/runtime/blob/5cd7e970487c4cc9ae652185a2f867159c48c9a0/src/coreclr/src/System.Private.CoreLib/src/System/RuntimeType.CoreCLR.cs#L1313
) and not returned. In `case1` the signature match but in `case2` the signatures are different.
